plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3Service.java

```diff
@@ -259,23 +278,91 @@ public Socket createSocket(final HttpContext ctx) throws IOException {
+            if (region != null || credentials != null) {
+                securityTokenService = SocketAccess.doPrivileged(
+                    () -> AWSSecurityTokenServiceClientBuilder.standard()
+                        .withCredentials((credentials != null) ? new AWSStaticCredentialsProvider(credentials) : null)
+                        .withRegion(region)
+                        .build()
+                );
+            }
+
+            if (irsaCredentials.getIdentityTokenFile() == null) {
+                return new PrivilegedSTSAssumeRoleSessionCredentialsProvider<>(
+                    securityTokenService,
+                    new STSAssumeRoleSessionCredentialsProvider.Builder(irsaCredentials.getRoleArn(), irsaCredentials.getRoleSessionName())
+                        .withStsClient(securityTokenService)
+                        .build()
+                );
+            } else {
+                return new PrivilegedSTSAssumeRoleSessionCredentialsProvider<>(
+                    securityTokenService,
+                    new STSAssumeRoleWithWebIdentitySessionCredentialsProvider.Builder(
+                        irsaCredentials.getRoleArn(),
+                        irsaCredentials.getRoleSessionName(),
+                        irsaCredentials.getIdentityTokenFile()
+                    ).withStsClient(securityTokenService).build()
+                );
+            }
+        } else if (credentials != null) {
             logger.debug("Using basic key/secret credentials");
             return new AWSStaticCredentialsProvider(credentials);
+        } else {
+            logger.debug("Using instance profile credentials");
+            return new PrivilegedInstanceProfileCredentialsProvider();
+        }
+    }
+
+    private static IrsaCredentials buildFromEnviroment(IrsaCredentials defaults) {
+        if (defaults == null) {
+            return null;
+        }
+
+        String webIdentityTokenFile = defaults.getIdentityTokenFile();
+        if (webIdentityTokenFile == null) {
+            webIdentityTokenFile = System.getenv(AWS_WEB_IDENTITY_ENV_VAR);
+        }
+
+        String roleArn = defaults.getRoleArn();
+        if (roleArn == null) {
+            roleArn = System.getenv(AWS_ROLE_ARN_ENV_VAR);
+        }
+
+        String roleSessionName = defaults.getRoleSessionName();
+        if (roleSessionName == null) {
+            roleSessionName = System.getenv(AWS_ROLE_SESSION_NAME_ENV_VAR);
         }
+
+        return new IrsaCredentials(webIdentityTokenFile, roleArn, roleSessionName);
     }
 
     private synchronized void releaseCachedClients() {
         // the clients will shutdown when they will not be used anymore
         for (final AmazonS3Reference clientReference : clientsCache.values()) {
             clientReference.decRef();
         }
+
+        for (final Closeable closeable : credentialsCache) {
+            try {
+                closeable.close();
```

Comment:
Is the safe? The reference counted clients use these credentials, and the clients can still be in use at this point, right?

As an alternative to tracking creds in a map in this class, there appears to be a 1-to-1 relationship between an AmazonS3Reference instance and a credentials instance. Did you consider extending AmazonS3Reference to own the credentials instance and close that when the client is ultimately closed? That would save you from needing to track it separately here.

Reply:
I think yeah, it will be safer to release credentials in lock step with the client, thank you

Reply:
@andrross updated + new tests added, thanks!